### PR TITLE
Subscriptions

### LINKED ISSUE #506: Accept bitcoin donations (coinbase)
Think we can just set up a coinbase wallet.  If we wanted to link into the merchant API we could also support cheese purchasing via bitcoin

### PR DESCRIPTION
**Pull request by albertcui**
Fixes #506, #388, #721.

### CODE REVIEW

File: public/js/donate.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+var $canceInfo = $("#cancel-info");
```
Comment (howardchung) [inline public/js/donate.js]: spelling?

File: routes/donate.js [hunk 2]
```diff
@@ -1,133 +1,280 @@
...
-                        "currency": "USD"
-                    },
-                    "description": "Buying CHEESE x" + num
-        }]
-            };
-            paypal.payment.create(payment, function(err, payment) {
-                if (err) {
-                    return next(err);
+        var amount = Number(req.body.amount);
+        var subscription = req.body.subscription != "false";
```
Comment (howardchung) [inline routes/donate.js]: !==

File: routes/donate.js [hunk 3]
```diff
@@ -1,133 +1,280 @@
...
+        var token = req.body.token;
+        
+        if (!token || amount == "NaN") {
```
Comment (howardchung) [inline routes/donate.js]: is isNaN for safe comparison

File: routes/donate.js [hunk 4]
```diff
@@ -1,133 +1,280 @@
...
+            return res.sendStatus(500);
```
Comment (howardchung) [inline routes/donate.js]: does this also end res?

File: routes/donate.js [hunk 5]
```diff
@@ -1,133 +1,280 @@
...
+                        active_until: moment().add(1, "M").format("YYYY-MM-DD")
+                    }).asCallback(function(err) {
+                        if (err) return res.send(checkErr());
+                        
+                        console.log("Added subscription for %s, cusomer id %s",
+                                    req.user.account_id,
+                                    customer.id);
+                        
+                        req.session.cheeseAmount = amount;
+                        req.session.subscription = 1; // Signed in
```
Comment (howardchung) [inline routes/donate.js]: why 1 and 2?

File: routes/donate.js [hunk 6]
```diff
@@ -1,133 +1,280 @@
...
-                        var link = payment.links[i];
-                        if (link.method === 'REDIRECT') {
-                            redirectUrl = link.href;
-                        }
-                    }
-                    res.redirect(redirectUrl);
+            })
+        
+        } else {
+            stripe.charges.create({
```
Comment (howardchung) [inline routes/donate.js]: difference between charges and customers?

File: routes/donate.js [hunk 7]
```diff
@@ -1,133 +1,280 @@
...
+                amount: amount * 100,
+                currency: "usd",
+                source: token.id,
+                description: "Buying " + amount + " cheese!"
+            }, function(err, charge) {
+                if (err) return res.send(checkErr(err));
```
Comment (howardchung) [inline routes/donate.js]: use {}

File: routes/donate.js [hunk 8]
```diff
@@ -1,133 +1,280 @@
...
-    donate.route('/cancel').get(function(req, res) {
-        clearPaymentSessions(req);
-        res.render("cancel");
+    donate.route("/cancel").get(function(req, res, next) {
+        if (!req.user) return res.render("cancel", {
+            sub: false
+        });
+        
+        db("subscriptions")
+        .where({
```
Comment (howardchung) [inline routes/donate.js]: first() or you'll get array?

File: routes/donate.js [hunk 9]
```diff
@@ -1,133 +1,280 @@
...
+    donate.route("/stripe_endpoint").post(function(req, res, next) {
+        var id = req.body.id;
+        console.log("Got a event from Stripe, id %s", id);
+        // Get the event from Stripe to verify
+        stripe.events.retrieve(id, function(err, event) {
+            if (err) return res.sendStatus(400);
+            
+            // Only care about charge succeeded or subscription ended
+            if (event.type !== "charge.succeeded" && event.type !== "customer.subscription.deleted") {
+                return res.sendStatus(200);
```
Comment (howardchung) [inline routes/donate.js]: dont send this until we do all dbops

File: routes/donate.js [hunk 10]
```diff
@@ -1,133 +1,280 @@
...
+            }
+            
+            // Check that we haven't seen this before
+            redis.lrange("stripe:events", 0, 1000, function(err, result) {
+                if (err) return res.sendStatus(400); // Redis is derping, have Stripe send back later
```
Comment (howardchung) [inline routes/donate.js]: 500?

File: routes/donate.js [hunk 11]
```diff
@@ -1,133 +1,280 @@
...
 
+                for (var e in result) {
+                    
+                    if (e == id) {
```
Comment (howardchung) [inline routes/donate.js]: ===

File: routes/donate.js [hunk 12]
```diff
@@ -1,133 +1,280 @@
...
+            // Only care about charge succeeded or subscription ended
+            if (event.type !== "charge.succeeded" && event.type !== "customer.subscription.deleted") {
+                return res.sendStatus(200);
+            }
+            
+            // Check that we haven't seen this before
+            redis.lrange("stripe:events", 0, 1000, function(err, result) {
+                if (err) return res.sendStatus(400); // Redis is derping, have Stripe send back later
 
+                for (var e in result) {
```
Comment (howardchung) [inline routes/donate.js]: isnt result a [] not {}

File: routes/donate.js [hunk 13]
```diff
@@ -1,133 +1,280 @@
...
+                
+                // New event
+                if (event.type === "charge.succeeded") {
+                    var amount = event.data.object.amount/100;
+                    
+                    console.log("Event %s: Charge succeeded for %s.", id, amount);
+                    
+                    // Update cheese goal
+                    redis.incrby("cheese_goal", amount, function(err, val) {
+                        if (!err && val == amount) {
```
Comment (howardchung) [inline routes/donate.js]: use ===

File: routes/donate.js [hunk 14]
```diff
@@ -1,133 +1,280 @@
...
+                }
+                
+                // New event
+                if (event.type === "charge.succeeded") {
+                    var amount = event.data.object.amount/100;
+                    
+                    console.log("Event %s: Charge succeeded for %s.", id, amount);
+                    
+                    // Update cheese goal
+                    redis.incrby("cheese_goal", amount, function(err, val) {
```
Comment (howardchung) [inline routes/donate.js]: do you just want to swallow redis exception?  at least log it

File: routes/donate.js [hunk 15]
```diff
@@ -1,133 +1,280 @@
...
+                            db('subscriptions')
+                            .returning("account_id")
+                            .update({
+                                active_until: moment().add(1, "M").format("YYYY-MM-DD")
+                            })
+                            .where({
+                                customer_id: customer
+                            })
+                            .asCallback(function(err, sub) {
+                                if (err) return res.sendStatus(400); // Postgres derping
```
Comment (howardchung) [inline routes/donate.js]: 500?

File: routes/donate.js [hunk 16]
```diff
@@ -1,133 +1,280 @@
...
+                                    addEventAndRespond(id, res);
+                                }
+                            })
+                        } else {
+                            addEventAndRespond(id, res);
+                        }
+                    })
+                } else if (event.type === "customer.subscription.deleted") {
+                    // Our delete process should delete the subscription, but make sure.
+                    var customer = event.data.object.customer
```
Comment (howardchung) [inline routes/donate.js]: semicolon

File: routes/donate.js [hunk 17]
```diff
@@ -1,133 +1,280 @@
...
+                    console.log("Event %s: Customer %s being deleted.", id, customer);
+                    
+                    db("subscriptions")
+                    .where({
+                        customer_id: customer
+                    }).del()
+                    .asCallback(function(err, result) {
+                        if (err) return res.sendStatus(400);
```
Comment (howardchung) [inline routes/donate.js]: 500?  maybe make function for this

File: views/carry.jade [hunk 18]
```diff
@@ -4,36 +4,71 @@ block content
...
-                option 100
-              p CHEESE
-          div
-            button.btn.btn-warning(type="submit") CHEESE IT
-        div(style="margin-top:1em;")
-          em $1/Cheese. It's worth it. Clicking the button will take you to Paypal.
+      if subscription && subscription.length > 0
+        .row
+          .col-md-8.col-md-offset-2
+            if subscription.length > 1
```
Comment (howardchung) [inline views/carry.jade]: cant you unify n=1 and n > 1?

File: routes/donate.js (hunk 4 above)
Comment (albertcui) [inline routes/donate.js]: yes

File: routes/donate.js (hunk 5 above)
Comment (albertcui) [inline routes/donate.js]: signed in or not signed in

File: routes/donate.js (hunk 6 above)
Comment (albertcui) [inline routes/donate.js]: customers are subscriptions. charges are one time.

File: routes/donate.js (hunk 10 above)
Comment (albertcui) [inline routes/donate.js]: doesn't really matter, as long as not 2xx.

File: routes/donate.js (hunk 13 above)
Comment (albertcui) [inline routes/donate.js]: In the original subscriptions code this was == because redis gives a string back. I'm pretty sure this is still the case.

Comment (howardchung) [inline routes/donate.js]: Always use === and do the conversion yourself: Number(amount).

**Pull request merged**